cassandra40/src/main/java/org/apache/cassandra/sidecar/cassandra40/TokenRangeReplicas.java

```diff
@@ -0,0 +1,417 @@
+        {
+            return false;
+        }
+        TokenRangeReplicas that = (TokenRangeReplicas) o;
+        return (this.start.equals(that.start) && this.end.equals(that.end) && this.partitioner == that.partitioner);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int hashCode()
+    {
+        return Objects.hashCode(start, end, partitioner);
+    }
+
+    private boolean isWrapAround()
+    {
+        return start.compareTo(end) >= 0;
+    }
+
+    private boolean isSubsetOf(TokenRangeReplicas other)
+    {
+        if (this.partitioner != other.partitioner)
+            throw new IllegalStateException("Token ranges being compared do not have the same partitioner");
+
+        BigInteger maxValue = this.partitioner.maxToken;
+        if (this.start.compareTo(other.start) >= 0)
+        {
+            // TODO:
+            if (other.end.equals(maxValue)) return true;
+            if (this.end.equals(maxValue)) return false;
+            if (this.end.compareTo(other.end) <= 0) return true;
+        }
+        return false;
+    }
+
+    /**
+     * For subset ranges, this is used to determine if a range is larger than the other by comparing start-end lengths
+     * If both ranges end at the min, we compare starting points to determine the result.
+     * When the left range is the only one ending at min, it is always the larger one since all subsequent ranges
+     * in the sorted range list have to be smaller.
+     *
+     * @param other the next range in the range list to compare
+     * @return true if "this" range is larger than the other
+     */
+    private boolean isLarger(TokenRangeReplicas other)
+    {
+        if (this.partitioner != other.partitioner)
+            throw new IllegalStateException("Token ranges being compared do not have the same partitioner");
+
+        // If both ranges end at min, we compare start of ranges
+        if (this.end.equals(partitioner.maxToken) && other.end.equals(partitioner.maxToken))
+        {
+            return this.start.compareTo(other.start) < 0;
+        }
+
+        if (this.end.equals(partitioner.maxToken)) return true;
+        if (other.end.equals(partitioner.maxToken)) return false;
+
+        return this.end.subtract(this.start).compareTo(other.end.subtract(other.start)) > 0;
+    }
+
+    /**
+     * Determines intersection if the next range starts before the current range ends.
+     * When the current range ending at min, we determine intersection merely if the next range starts after the current
+     * since all subsequent ranges have to be subsets.
+     *
+     * @param other the range we are currently processing to check if "this" intersects it
+     * @return true if "this" range intersects the other
+     */
+    private boolean intersects(TokenRangeReplicas other)
+    {
+        return (other.end.compareTo(partitioner.maxToken) == 0 && this.start.compareTo(other.start) > 0) ||
+               this.start.compareTo(other.end) < 0;
+    }
```

Comment:
I think this implementation is wrong. You want to compare `this.end.compareTo(other.start) < 0;`. Before the comparison, it needs to sort the 2 ranges by `start`, and make sure to compare left range with right range.
Beside that, the first condition that checks with the maxToken of the partitioner seems not necessary. It only wants to compare 2 ranges.  

The below test fails 

```java
    @Test
    void testInterests()
    {
        TokenRangeReplicas range1 = new TokenRangeReplicas(BigInteger.valueOf(1), BigInteger.valueOf(10), Partitioner.Murmur3, new HashSet<>());
        TokenRangeReplicas range2 = new TokenRangeReplicas(BigInteger.valueOf(9), BigInteger.valueOf(12), Partitioner.Murmur3, new HashSet<>());
        assertThat(range1.intersects(range2)).isTrue();
        assertThat(range2.intersects(range1)).isTrue();

        TokenRangeReplicas range3 = new TokenRangeReplicas(BigInteger.valueOf(1), BigInteger.valueOf(10), Partitioner.Murmur3, new HashSet<>());
        TokenRangeReplicas range4 = new TokenRangeReplicas(BigInteger.valueOf(11), BigInteger.valueOf(20), Partitioner.Murmur3, new HashSet<>());
        assertThat(range3.intersects(range4)).isFalse();
        assertThat(range4.intersects(range3)).isFalse();
    }
```